asnake/aspace/__init__.py

```diff
@@ -0,0 +1,138 @@
+				#for calls without repositories in them
+				if attr in shortCalls:
+					return jsonmodel_muliple_object(self.__client.get("/" + str(attr) + "?all_ids=true").json(), self.__client, self.repository, attr)
+				else:
+					return jsonmodel_muliple_object(self.__client.get("/repositories/" + str(self.repository) + "/" + str(attr) + "?all_ids=true").json(), self.__client, self.repository, attr)
+
+	# not sure if theres a way to pass a variable to implement this with __getattr__
+	def resource(self, id):
+		return jsonmodel_single_object(self.__client.get("repositories/" + self.repository + "/resources/" + str(id)).json(), self.__client)
+	
+	#this doesn't work yet
+	def resourceID(self, id):
+		result = self.__client.get("/repositories/" + self.repository + "/search?page=1&aq={\"query\":{\"field\":\"identifier\", \"value\":\"" + str(id) + "\", \"jsonmodel_type\":\"field_query\"}}").json()
+		resourceURI = result["results"][0]["uri"]
+		return jsonmodel_single_object(self.__client.get(resourceURI).json(), self.__client)		
+		
+	def archival_object(self, id):
+		if isinstance(id, str):
+			if len(id) == 32:
+				# its a ref_id
+				params = {"ref_id[]": str(id)}
+				refList = self.__client.get("repositories/" + self.repository + "/find_by_id/archival_objects?page=1&ref_id[]=" + str(id)).json()
+				return jsonmodel_single_object(self.__client.get(refList["archival_objects"][0]["ref"]).json(), self.__client)
+		#its a uri number
+		return jsonmodel_single_object(self.__client.get("repositories/" + self.repository + "/archival_objects/" + str(id)).json(), self.__client)
+	
+	def agents(self, type, id = None):
+		if id == None:
+			return jsonmodel_muliple_object(self.__client.get("/agents/" + str(type) + "?all_ids=true").json(), self.__client, self.repository, type)
+		else:
+			return jsonmodel_single_object(self.__client.get("/agents/" + str(type) + "/" + str(id)).json(), self.__client)
+
+		
+class jsonmodel_single_object:
+	def __init__(self, json_rep, client = None):
+		self.__json = json_rep
+		self.__client = client
+	
+	def __getattr__(self, key):
+	
+		if isinstance(key, str):
+			if not key.startswith('_'):
+				if not key in self.__json.keys():
+					repository = self.__json["repository"]["ref"].split("/repositories/")[1]
+					return jsonmodel_muliple_object(self.__json, self.__client)
+		
+		if isinstance(self.__json[key], list):
+			return jsonmodel_muliple_object(self.__json[key], self.__client)
+		elif isinstance(self.__json[key], dict):
+			return jsonmodel_single_object(self.__json[key], self.__client)
+		else:
+			return self.__json[key]
+
+	# utility methods
+	def pp(self):
+		print (json.dumps(self.__json, indent=2))
+
+	def json(self):
+		return self.__json
+
+	def serialize(self, filePath):
+		f = open(filePath, "w")
+		f.write(json.dumps(self.__json, indent=2))
+		f.close
+
+#Returns a generator object to stream lists of objects
+def jsonmodel_muliple_object(json_list, client, repository=None, call=None):
```

Comment:
This is v good, and how I initially wanted to implement it - Tristan suggested (and I think is probably right) that it might be faster when doing the multiple object stuff to do paging (keeping the current page and index on page in the generator as local vars) rather than all_ids; I'm not sure if it really matters enough to change, though.  It would also let us implement len() basically for free, since the paged views contain that information.

Reply:
I think a way that might make sense is to have the jsonmodel_multiple_object be a class rather than a method, in its __init__ fetch the first page, offer some methods to get length, etc, and then implement __iter__ to be paging over the members of the collection (and maybe a get method or __getitem__ as "fetch the one with this id"?)

Reply:
So I implemented paging and stuff in that PR I made to _your_ branch, but wanted to just say that we actually _don't_ get count for free bc count isn't included in paged results, just current page, last page.  It doesn't even include page_size, which is egregious IMO and I'm gonna file a bug report, since, well, it means you can't safely reason about paging if you didn't explicitly set a page size (I think the default is ten, but it isn't guaranteed not to change and it's not mentioned in docs that I found.

Reply:
Yeah, its probably better as a class, I was just learning from an example and didn't think to put it in init(), I'll take a look at what did, seems like paginating might also let us optionally loop my page too.